### PR TITLE
Fix dmg notarization

### LINKED ISSUE #131: Mac Codesigning and dmg notarization not working
Hi,
I've run into a case where my electron application crashes due to invalid signing of some files on mac. This mainly applies to some native binaries (dylib and executables) inside the bundled framework directories. This also breaks the notarization process.

#### 1. "deep" codesigning will break complex applications in the hardened runtime
[As described by Apple](https://developer.apple.com/forums/thread/129980) it is highly discouraged to use the `--deep` codesign flag. I therefore suggest manually signing all relevant files (which should be mostly Mach-O files).

#### 2. Timestamps are missing
When submitting an application for notarization, Apple requires a timestamp metadata to indicate when the file was signed.
`--timestamp` should be used on every codesign process.

### PR DESCRIPTION
**Pull request by jan-br**
Closes #131

### CODE REVIEW

**Comment by denjell-crabnebula** [comment]
Thanks for the PR, we will review it today!

File: crates/packager/src/package/app/mod.rs [hunk 1]
```diff
@@ -52,6 +52,77 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
+    // https://developer.apple.com/forums/thread/129980
+    for framework_path in &framework_paths {
+        if let Some(framework_path) = framework_path.to_str() {
+            // Find all files in the current framework folder
+            let constructed_glob = format!("{}/**/*", framework_path);
+            let files = match glob::glob(&constructed_glob) {
+                Ok(res) => res,
+                Err(err) => {
+                    tracing::warn!("Could not construct a glob for {}: {}, please make sure your path contains no *", framework_path, err);
+                    tracing::warn!("This path will not be scanned for dylibs!");
```
Comment (amr-crabnebula) [inline crates/packager/src/package/app/mod.rs:67]: Let's merge these into a single `warn!` call

File: crates/packager/src/package/app/mod.rs [hunk 2]
```diff
@@ -52,6 +52,77 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
+                Ok(v) => Some(v),
+                Err(err) => {
+                    tracing::warn!(
+                        "Error while globbing for Mach-O files in {}: {}",
+                        framework_path,
+                        err
+                    );
+                    None
+                }
+            });
```
Comment (amr-crabnebula) [inline crates/packager/src/package/app/mod.rs:82]: We can just use flatten, don't think we need log these errors, we can just skip them
```suggestion
            let files = files.flatten();
```

File: crates/packager/src/package/app/mod.rs [hunk 3]
```diff
@@ -52,6 +52,77 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
+
+            // Filter all files for Mach-O headers. This will target all .dylib and native executable files
+            for file in files {
+                let Some(file) = file.to_str() else {
+                    tracing::warn!("Failed to convert {} to a valid UTF-8 string, this path will not be scanned for Mach-O header!", file.display());
+                    continue;
+                };
```
Comment (amr-crabnebula) [inline crates/packager/src/package/app/mod.rs:89]: this is not needed, since `fs::metadata` can accept `PathBuf` just fine

**Comment by denjell-crabnebula** [comment]
Hi @jan-br - would you mind also adding a new [.changes](https://github.com/crabnebula-dev/cargo-packager/blob/main/.changes) file like this one:

https://github.com/crabnebula-dev/cargo-packager/blob/main/.changes/priority-and-section.md

File: crates/packager/src/package/app/mod.rs [hunk 4]
```diff
@@ -52,6 +52,77 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
+
+                let Ok(metadata) = std::fs::metadata(file) else {
```
Comment (amr-crabnebula) [inline crates/packager/src/package/app/mod.rs:91]: we should match here and append the error to the log message 
```rs
let metadata = match std::fs::metadata(file) {
	Ok(f) => f,
	Err(err) => {
	  tracing::warn!("Failed to get metadata for {}: {err}, this file will not be scanned for Mach-O header!", file);
      continue;
	}
};
```

File: crates/packager/src/package/app/mod.rs [hunk 5]
```diff
@@ -52,6 +52,77 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
+                };
+
+                if !metadata.is_file() {
+                    continue;
+                }
+
+                let Ok(mut open_file) = std::fs::File::open(file) else {
+                    tracing::warn!("Failed to open {} for reading, this file will not be scanned for Mach-O header!", file);
+                    continue;
+                };
```
Comment (amr-crabnebula) [inline crates/packager/src/package/app/mod.rs:103]: same here

File: crates/packager/src/package/app/mod.rs [hunk 6]
```diff
@@ -52,6 +52,77 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
+
+                let mut buffer = [0; 4];
+                std::io::Read::read_exact(&mut open_file, &mut buffer)?;
+
+                // Mach-O magic numbers
+                let magic_numbers = [0xfeedface, 0xfeedfacf, 0xcafebabe, 0xcefaedfe, 0xcffaedfe];
```
Comment (amr-crabnebula) [inline crates/packager/src/package/app/mod.rs:109]: let's use a const here
```suggestion
                const MACH_O_MAGIC_NUMBERS: [u32; 5] = [0xfeedface, 0xfeedfacf, 0xcafebabe, 0xcefaedfe, 0xcffaedfe];
```

**Review by amr-crabnebula — CHANGES_REQUESTED** [review]

**Comment by jan-br** [comment]
Hi @amr-crabnebula @denjell-crabnebula,
thanks for the feedback. Does this look good to you now?

File: crates/packager/src/package/app/mod.rs [hunk 7]
```diff
@@ -52,6 +52,65 @@ pub(crate) fn package(ctx: &Context) -> crate::Result<Vec<PathBuf>> {
...
     tracing::debug!("Copying frameworks");
     let framework_paths = copy_frameworks_to_bundle(&contents_directory, config)?;
+
+    // All dylib files and native executables should be signed manually
+    // It is highly discouraged by Apple to use the --deep codesign parameter in larger projects.
+    // https://developer.apple.com/forums/thread/129980
+    for framework_path in &framework_paths {
+        if let Some(framework_path) = framework_path.to_str() {
+            // Find all files in the current framework folder
+            let constructed_glob = format!("{}/**/*", framework_path);
```
Comment (lucasfernog-crabnebula) [inline crates/packager/src/package/app/mod.rs:62]: isn't it better to use [walkdir](https://docs.rs/walkdir/latest/walkdir/)?

Comment (jan-br) [reply crates/packager/src/package/app/mod.rs:62]: Ah yea, I didn't see that you already have a dependency on walkdir and didn't want to add one. I'll change that

**Comment by jan-br** [comment]
Anything else @lucasfernog-crabnebula?

**Review by amr-crabnebula — APPROVED** [review]

**Comment by amr-crabnebula** [comment]
looks like you need to run `pnpm format` and `cargo fmt`, don't worry about deny workflow failing

**Review by amr-crabnebula — APPROVED** [review]

**Pull request merged**